services/director-v2/src/simcore_service_director_v2/cli/core.py

```diff
@@ -0,0 +1,298 @@
+from tenacity.wait import wait_random_exponential
+
+from ..core.application import create_base_app
+from ..core.settings import AppSettings
+from ..models.domains.dynamic_services import DynamicServiceOut
+from ..models.schemas.dynamic_services import (
+    DynamicSidecarNamesHelper,
+    ServiceBootType,
+    ServiceState,
+)
+from ..modules import db, director_v0, dynamic_sidecar
+from ..modules.db.repositories.projects import ProjectsRepository
+from ..modules.director_v0 import DirectorV0Client
+from ..modules.dynamic_sidecar import api_client
+from ..modules.dynamic_sidecar.scheduler._utils import get_repository
+from ..modules.projects_networks import requires_dynamic_sidecar
+
+
+@asynccontextmanager
+async def _initialized_app(only_db: bool = False) -> AsyncIterator[FastAPI]:
+    app = create_base_app()
+    settings: AppSettings = app.state.settings
+
+    # Initialize minimal required components for the application
+    db.setup(app, settings.POSTGRES)
+
+    if not only_db:
+        dynamic_sidecar.setup(app)
+        director_v0.setup(app, settings.DIRECTOR_V0)
+
+    await app.router.startup()
+    yield app
+    await app.router.shutdown()
+
+
+### PROJECT SAVE STATE
+
+
+def _get_dynamic_sidecar_endpoint(
+    settings: AppSettings, node_id: NodeIDStr
+) -> AnyHttpUrl:
+    dynamic_sidecar_names = DynamicSidecarNamesHelper.make(UUID(node_id))
+    hostname = dynamic_sidecar_names.service_name_dynamic_sidecar
+    port = settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR.DYNAMIC_SIDECAR_PORT
+    return parse_obj_as(AnyHttpUrl, f"http://{hostname}:{port}")  # NOSONAR
+
+
+async def _save_node_state(
+    app,
+    dynamic_sidecar_client: api_client.DynamicSidecarClient,
+    retry_save: int,
+    node_uuid: NodeIDStr,
+    label: str = "",
+) -> None:
+    typer.echo(f"Saving state for {node_uuid} {label}")
+    async for attempt in AsyncRetrying(
+        wait=wait_random_exponential(),
+        stop=stop_after_attempt(retry_save),
+        reraise=True,
+    ):
+        with attempt:
+            typer.echo(f"Attempting to save {node_uuid} {label}")
+            await dynamic_sidecar_client.save_service_state(
+                _get_dynamic_sidecar_endpoint(app.state.settings, node_uuid)
+            )
+
+
+async def async_project_save_state(project_id: ProjectID, retry_save: int) -> None:
+    async with _initialized_app() as app:
+        projects_repository: ProjectsRepository = get_repository(
+            app, ProjectsRepository
+        )
+        project_at_db = await projects_repository.get_project(project_id)
+
+        typer.echo(f"Saving project '{project_at_db.uuid}' - '{project_at_db.name}'")
+
+        dynamic_sidecar_client = api_client.get_dynamic_sidecar_client(app)
+        nodes_failed_to_save: list[NodeIDStr] = []
+        for node_uuid, node_content in project_at_db.workbench.items():
+            # onl dynamic-sidecars are used
+            if not await requires_dynamic_sidecar(
```

Comment:
Q: do we have studies with legacy services? if yes what of them?

Reply:
Yes we still have some of there here and there. They are being deprecated at the moment, so I expect them to be rolled out.